src/Data/Vector/Hashtables/Internal.hs

```diff
@@ -541,6 +541,48 @@ findWithDefault
 findWithDefault ht v k = return . fromMaybe v =<< at' ht k
 {-# INLINE findWithDefault #-}
 
+-- | /O(1)/ in the best case, /O(n)/ in the worst case.
+-- The expression (@'upsert' ht f k@) updates or inserts the value @x@ at @k@.
+--
+-- > let f _ = "c"
+-- > ht <- fromList [(5,"a"), (3,"b")]
+-- > upsert ht f 7
+-- > toList ht
+-- > [(3, "b"), (5, "a"), (7, "c")]
+--
+-- > ht <- fromList [(5,"a"), (3,"b")]
+-- > upsert ht f 5
+-- > toList ht
+-- > [(3, "b"), (5, "c")]
+--
+upsert
+  :: ( MVector ks k, MVector vs v
+     , PrimMonad m, Hashable k, Eq k
+     )
+  => Dictionary (PrimState m) ks k vs v -> (Maybe v -> v) -> k -> m ()
+upsert ht f k = do
+  d@Dictionary{..} <- readMutVar . getDRef $ ht
+  let
+      hashCode' = hash k .&. mask
+      targetBucket = hashCode' `rem` A.length buckets
+
+      onFound' value' dict i = insertWithIndex targetBucket hashCode' k value' (getDRef ht) dict i
+
+      onFound dict i = do
+        d'@Dictionary{..} <- readMutVar . getDRef $ dict
+        v <- value !~ i
+        case f (Just v) of
+          v' ->  onFound' v' d' i
```

Comment:
Why do you need `case` here? Plain `onFound' (Just v) d' i` would be semantically identical as `case`s in the surface Haskell don't evaluate anything (as opposed to GHC Core).

Same below.

Reply:
Whatever you prefer, updated.

Reply:
Thanks!